flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/nodes/physical/stream/StreamExecTemporalJoin.scala

```diff
@@ -84,6 +101,361 @@ class StreamExecTemporalJoin(
+      if (TemporalJoinUtil.isRowTimeJoin(rexBuilder, joinInfo)) {
+        checkState(temporalJoinConditionExtractor.leftTimeAttribute.isDefined &&
+          temporalJoinConditionExtractor.rightPrimaryKey.isDefined,
+          "Missing %s in Event-Time temporal join condition", TEMPORAL_JOIN_CONDITION)
+
+        val leftTimeAttributeInputRef = extractInputRef(
+          temporalJoinConditionExtractor.leftTimeAttribute.get, textualRepresentation)
+        val rightTimeAttributeInputRef = extractInputRef(
+          temporalJoinConditionExtractor.rightTimeAttribute.get, textualRepresentation)
+        val rightInputRef = rightTimeAttributeInputRef - leftType.getFieldCount
+
+        (leftTimeAttributeInputRef, Some(rightInputRef))
+      } else {
+        val leftTimeAttributeInputRef = extractInputRef(
+          temporalJoinConditionExtractor.leftTimeAttribute.get, textualRepresentation)
+        // right time attribute defined in temporal join condition iff in Event time join
+        (leftTimeAttributeInputRef, None)
+      }
+
+    new StreamExecTemporalJoinToCoProcessTranslator(
+      textualRepresentation,
+      config,
+      returnType,
+      leftType,
+      rightType,
+      joinInfo,
+      rexBuilder,
+      leftTimeAttributeInputRef,
+      rightRowTimeAttributeInputRef,
+      remainingNonEquiJoinPredicates,
+      isTemporalFunctionJoin)
+  }
+
+  private def extractInputRef(rexNode: RexNode, textualRepresentation: String): Int = {
+    val inputReferenceVisitor = new InputRefVisitor
+    rexNode.accept(inputReferenceVisitor)
+    checkState(
+      inputReferenceVisitor.getFields.length == 1,
+      "Failed to find input reference in [%s]",
+      textualRepresentation)
+    inputReferenceVisitor.getFields.head
+  }
+
+  private class TemporalJoinConditionExtractor(
+    textualRepresentation: String,
+    rightKeysStartingOffset: Int,
+    joinInfo: JoinInfo,
+    rexBuilder: RexBuilder,
+    isTemporalFunctionJoin: Boolean) extends RexShuttle {
+
+    var leftTimeAttribute: Option[RexNode] = None
+
+    var rightTimeAttribute: Option[RexNode] = None
+
+    var rightPrimaryKey: Option[Array[RexNode]] = None
+
+    override def visitCall(call: RexCall): RexNode = {
+      if (call.getOperator != TEMPORAL_JOIN_CONDITION) {
+        return super.visitCall(call)
+      }
+      // the condition of temporal function comes from WHERE clause,
+      // so it's not been validated in logical plan
+      if (isTemporalFunctionJoin) {
+        validateAndExtractTemporalFunctionCondition(call)
+      }
+      // temporal table join has been validated in logical plan, only do extract here
+      else {
+        if (TemporalJoinUtil.isRowTimeTemporalTableJoinCon(call)) {
+          leftTimeAttribute = Some(call.getOperands.get(0))
+          rightTimeAttribute = Some(call.getOperands.get(1))
+          rightPrimaryKey = Some(extractPrimaryKeyArray(call.getOperands.get(4)))
```

Comment:
We can put the primary key at the 2nd position, then the extraction logic should be the same with legacy temporal join.